ua_parser/__init__.py

```diff
@@ -1 +1,201 @@
+
+
+@dataclass
+class Device:
+    name: str
+    brand: Optional[str] = None
+    model: Optional[str] = None
+
+    def __str__(self):
+        """Would replace PrettyDevice if that was a thing"""
+        ...
+
+
+class RegexDeviceMatcher:
+    def __call__(self, uas: str) -> Optional[Device]:
+        ...
+
+
+T = TypeVar("T")
+# TODO: checking types: https://sobolevn.me/2019/08/testing-mypy-types
+Matcher = Callable[[str], Optional[T]]
+Matchers = Tuple[
+    Iterable[Matcher[UserAgent]],
+    Iterable[Matcher[OS]],
+    Iterable[Matcher[Device]],
+]
+
+
+@dataclass
+class ParseResult:
+    user_agent: UserAgent
+    os: OS
+    device: Device
+    string: str
+
+
+@dataclass
+class CacheEntry:
+    string: str
+    user_agent: Optional[UserAgent] = None
+    os: Optional[OS] = None
+    device: Optional[Device] = None
+
+    def to_result(self) -> Optional[ParseResult]:
+        if self.user_agent and self.os and self.device:
+            return ParseResult(self.user_agent, self.os, self.device, self.string)
+        return None
+
+
+# FIXME: should probably be a typing_extensions.Protocol with just
+#        "lookup", with this as a base helper
+class Cache(abc.ABC):
+    @abc.abstractmethod
+    def get(self, key) -> Optional[CacheEntry]:
+        ...
+
+    @abc.abstractmethod
+    def store(self, key, CacheEntry) -> CacheEntry:
+        ...
+
+    @abc.abstractmethod
+    def make_room(self):
+        ...
+
+    @property
+    @abc.abstractmethod
+    def maxsize(self) -> int:
+        ...
+
+    @abc.abstractmethod
+    def __len__(self) -> int:
+        ...
+
+    def check_space(self):
+        if len(self) >= self.maxsize:
+            self.make_room()
+
+    def key(
+        self, ua: str, args: dict
+    ):  # TODO: keys are internal cache stuff, how can they type?
+        return (ua, tuple(sorted(args.items())))
+
+    def lookup(self, ua: str, args: dict) -> CacheEntry:
+        key = self.key(ua, args)
+        entry = self.get(key)
+        if entry:
+            return entry
+
+        self.check_space()
+        return self.store(key, CacheEntry(string=ua))
+
+
+class Parser:
+    def __init__(
+        self,
+        *matchers: Tuple[Union[Matchers, str, IO]],
+        cache: Optional[Cache] = None,
+    ):
+        ms: Matchers
+        if not matchers:
+            from . import _regexes
+
+            ms = (
+                _regexes.USER_AGENT_PARSERS,
+                _regexes.OS_PARSERS,
+                _regexes.DEVICE_PARSERS,
+            )
+        elif not isinstance(matchers[0], tuple):
+            ms = Parser.yaml_to_matchers(matchers)
+        else:
+            ms = cast(Matchers, matchers[0])
+
+        self._ua_matchers, self._os_matchers, self._device_matchers = ms
+        self.cache = (
+            cache or ...
+        )  # should default to something similar to current, possibly via function
+
+    @staticmethod
+    def yaml_to_matchers(yaml_file: Union[str, IO]) -> "Matchers":
+        ...
```

Comment:
Probably makes more sense for this to be in a utility module dedicated to yaml, or maybe just a utility function alongside the "default" matchers.